cmd/snap-confine/mount-support.c

```diff
@@ -308,75 +508,29 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 	// mounted anywhere. When we construct recursive bind mounts below this
 	// guarantees that this directory will not be replicated anywhere.
 	sc_do_mount("none", scratch_dir, NULL, MS_UNBINDABLE, NULL);
-	// Recursively bind mount desired root filesystem directory over the
-	// scratch directory. This puts the initial content into the scratch space
-	// and serves as a foundation for all subsequent operations below.
-	//
-	// The mount is recursive because it can either be applied to the root
-	// filesystem of a core system (aka all-snap) or the core snap on a classic
-	// system. In the former case we need recursive bind mounts to accurately
-	// replicate the state of the root filesystem into the scratch directory.
-	sc_do_mount(config->rootfs_dir, scratch_dir, NULL, MS_REC | MS_BIND,
-		    NULL);
+	if (config->normal_mode) {
+		// Create a tmpfs on scratch_dir; we'll them mount all the root
+		// directories of the base snap onto it.
+		sc_do_mount("none", scratch_dir, "tmpfs", 0, NULL);
+		sc_replicate_rootfs(scratch_dir, config->rootfs_dir, config->mounts);
+	} else {
+		// Recursively bind mount desired root filesystem directory over the
+		// scratch directory. This puts the initial content into the scratch
+		// space and serves as a foundation for all subsequent operations
+		// below.
+		//
+		// The mount is recursive because we need to accurately replicate the
+		// state of the root filesystem into the scratch directory.
+		sc_do_mount(config->rootfs_dir, scratch_dir, NULL, MS_REC | MS_BIND,
+		            NULL);
+	}
 	// Make the scratch directory recursively slave. Nothing done there will be
 	// shared with the initial mount namespace. This effectively detaches us,
 	// in one way, from the original namespace and coupled with pivot_root
 	// below serves as the foundation of the mount sandbox.
 	sc_do_mount("none", scratch_dir, NULL, MS_REC | MS_SLAVE, NULL);
```

Comment:
This is fine like it is. But I feel like in case `config->normal_mode`, we did not need `MS_REC` since we have already bind mounted directories as such. It do not know if it would actually affect speed. But maybe we can move this into the previous `if`/`else` to not have `MS_REC` in the `if` part.

Reply:
Also this, do you mean that we already do recursively mount on top of scratch_dir in the previous if/else and thus we should not need it here?

Reply:
I do not think my comment makes sense anymore. I think the code has been changed since in a way we should always use `MS_REC` here.